### PR TITLE
IBBoard/hotfix/twitter and mastodon session fixes

### PR DESCRIPTION
**Pull request by IBBoard**
I don't know if this is all how you wanted it to go, but I wanted to start adding features and the newest branch seemed to be incomplete, so I've fixed up some bits.

Key bits to note:
* Twitter and Mastodon auth seems to work (I've set up both accounts - but passed the URL back manually from the command-line, because I'm not installing the app yet)
* There's an ugly "fix" in to avoid a CRITICAL from librest when it tries to check if our non-expiring offline access token has expired
* There's another ugly "fix" where we have a hard-coded app password for the client secret, since librest doesn't like the argument being null - at the very least, we need a strong password generated there
* Client creation didn't want to have a constructor, even though other `SingleInstance` classes do. I wonder whether it's because it's also split as a partial?
* I've had to change a couple of bits to a less pretty implementation because the nice Vala code (like the `find` function with a lambda) were giving segfaults and other issues because of how the C code was created
* I wanted to work out why Mastodon has to use the code, but I've not worked it out yet

.

[Edit] ~~Actually... I've not got Twitter working yet. I had it working, but then I tried to fix Mastodon, and now I've only got my Mastodon account and can't add Twitter!~~

~~But they have both been independently working as read-only accounts!~~

[Edit] Working now. Apparently the examples in the Vala docs show behaviour that you can no longer rely on!

### CODE REVIEW

**Comment by CodedOre** [comment]
> I don't know if this is all how you wanted it to go, but I wanted to start adding features and the newest branch seemed to be incomplete, so I've fixed up some bits.

I didn't looked at the project for while. IIRC, the ’backend-session‘ should have anything, aside of the rework of the authentication. I had an "finished" branch which added this, but I wasn't really happy with the code, so I wanted to see if I can rework it, but I had to prioritise university then. I'm also not at home currently, so I cant push this branch right now.

I'll look into the PR in detail later on.

**Comment by IBBoard** [comment]
I did wonder whether you might have some more code that  you hadn't committed, but it had been a few months and this seemed like a way for me to get into the code again and work out how you were structuring it 🙂

I started off fixing CRITICALs on the main branch, before realising that this branch had 100+ more commits.

**Comment by IBBoard** [comment]
I just changed "password123" to "password12" in the Twitter server client secret and it still authenticated. I don't know whether that means I missed something, or whether it means that the client secret isn't used by Twitter and we could put any non-null value in there.

**Comment by CodedOre** [comment]
> I just changed "password123" to "password12" in the Twitter server client secret and it still authenticated. I don't know whether that means I missed something, or whether it means that the client secret isn't used by Twitter and we could put any non-null value in there.

Well, for native applications, client secrets should no longer be a thing. To quote [_OAuth 2 Simplified_](https://www.oauth.com/oauth2-servers/client-registration/client-id-secret/):

> If the developer is creating a “public” app (a mobile or single-page app), then you should not issue a client_secret to the app at all. This is the only way to ensure the developer won’t accidentally include it in their application. If it doesn’t exist, it can’t be leaked!

So, ideally Twitter should ignore it, and we can set it to an empty string.

**Comment by IBBoard** [comment]
The bit that confuses me is that the "Client Secret" section suggests that there is still a per-app password (hence why I supplied _something_) but if I change it and it still works then is it used?

I assumed it was still needed, but defined with the server on a per-app basis rather than supplied within the binary (as we do at the moment).

**Comment by CodedOre** [comment]
As far as I understand it, Client Secrets are still used when you develop an application that is running only on controlled hardware, like an server. But since you could extract secrets from an binary, it was deprecated for so called native applications in favor of the new authentication mechanism.

**Comment by CodedOre** [comment]
Just an status update: I'm not at home until the end of the week, so I will not review it until next week. Want to check how it looks compared to my current code first.

File: lib/Twitter/System/Session.vala [hunk 1]
```diff
@@ -56,6 +55,11 @@ public partial class Backend.Twitter.Session : AsyncInitable {
                                   server.client_secret,
                                   @"https://api.$(server.domain)/2/");
     proxy.access_token = access_token;
+    // Work around librest doing a DateTime comparison with the default non-expiring token
+    // 100 years should be more than enough for anyone!
+    var now = new GLib.DateTime.now();
+    var far_future = now.add_years (100);
+    proxy.set_expiration_date (far_future);
```
Comment (CodedOre) [inline lib/Twitter/System/Session.vala:62]: With Twitter's access token, I previously had the issue that the token would actually "expire" after some point. So, in the old `main` branch, I had code to [get a new access token](https://github.com/CodedOre/NewCaw/blob/67042def3b2b592dfede844588cfabcaa4d6d285/lib/Twitter/System/Account.vala#L156-L161) and [saves that](https://github.com/CodedOre/NewCaw/blob/67042def3b2b592dfede844588cfabcaa4d6d285/src/System/Session.vala#L108-L109) on each startup.

I thought that this was just a thing of the new API, but it seems this bit of code seems to do the trick as well!

File: src/Authentication/AuthView.vala [hunk 2]
```diff
@@ -126,6 +133,14 @@ public class AuthView : Gtk.Widget {
...
   }
 
+  /**
+   * Hides the browser page for when we want to go straight to OOB
+   */
+  public void skip_browser () {
+    browser_page.navigatable = false;
+    auth_leaflet.navigate (FORWARD);
+  }
+
```
Comment (CodedOre) [inline src/Authentication/AuthView.vala:143]: You mentioned in your [commit message](https://github.com/CodedOre/NewCaw/pull/60/commits/d66d4f651527e1fa02b33e9770365f60fa6ef53a) that on Hachyderm the `cawbird://` redirect does not work. It does work though on `mastodon.online` and `mastodon.social`. So that seems to be a instance-related issue, not a reason to skip that for all of Mastodon.

**Review by CodedOre — COMMENTED** [review]
There are a few points which needs discussion, but otherwise it looks good to me for what it does.

Since my last work on adding the authentication isn't complete, and you also have some fixes, I would say we eventually merge this instead of reviving my old branch.

File: lib/Twitter/Utils/TextUtils.vala [hunk 3]
```diff
@@ -65,7 +65,7 @@ namespace Backend.Twitter.Utils.TextUtils {
             var entity         = TextModule ();
             entity.type        = MENTION;
             entity.display     = "@" + obj.get_string_member ("username");
-            entity.target      = obj.get_string_member ("id");
+            entity.target      = obj.has_member ("id") ? obj.get_string_member ("id") : "by/username/" + obj.get_string_member ("username");
```
Comment (CodedOre) [inline lib/Twitter/Utils/TextUtils.vala:68]: I'm not sure that should stay as it is, because it would create an username based id in the cache.
However, I have no clue at all how exactly two-month-past myself wanted to develop this further as well...

tldr: I have to take a look at the entire mentions system...

File: lib/Base/System/Client.vala [hunk 4]
```diff
@@ -75,38 +77,38 @@ public partial class Backend.Client : Initable {
...
-      website: website,
-      redirect_uri: redirect_uri,
-      sessions: new SessionList (),
-      servers: new ServerList ()
-    );
+  public void configure (string id, string name, string website, string? redirect_uri = null, string? twitter_oauth_key = null) throws Error {
+    this.id = id;
+    this.name = name;
+    this.website = website;
+    this.redirect_uri = redirect_uri;
```
[the remark this answers is not in the document]
Comment (IBBoard) [reply lib/Base/System/Client.vala:105]: I meant to get back to this one. It didn't like having a `public Client` constructor. It said one already existed. At the time I couldn't work it out and so switched to the `configure` method. I only worked out the split-class design later and meant to check whether there was a constructor in the other file.

File: lib/Base/System/ClientState.vala [hunk 5]
```diff
@@ -361,6 +362,11 @@ public partial class Backend.Client : Object {
     }
   }
 
+  public void register_session(Session session) throws Error {
+    sessions.add(session);
+    store_state ();
+  }
```
[the remark this answers is not in the document]
Comment (IBBoard) [reply lib/Base/System/ClientState.vala:368]: This was replacing a plain `add()` elsewhere in the code. We could call `store_state` from outside, but I like the pattern of having a function to do it for us so that we don't forget if we have another location that adds sessions.

File: lib/Twitter/System/Session.vala (hunk 1 above)
Comment (IBBoard) [reply lib/Twitter/System/Session.vala:62]: I'm not sure. I keep getting a "Forbidden" every so often, and then Cawbird crashes and loses my Twitter account. I only noticed it after making this PR, so I've not looked in to it yet.

The date won't fix the Twitter expiry problem. It's purely because `librest` tries to compare the timestamps and the token doesn't have an expiration date set in the property. I thought I'd read something about the token being "perpetual" or "non-expiring", so I assumed it didn't have a date and `librest` was just failing to handle that. But I can't find where I read it. And other authentication issues suggest that that isn't the case.

File: src/Authentication/AuthView.vala (hunk 2 above)
Comment (IBBoard) [reply src/Authentication/AuthView.vala:143]: We need a button to bypass it, then. Or I need to install Cawbird (rather than running a dev build from `build/`) and see whether it behaves differently when I have the protocol handler installed. But the console for Twitter definitely prints a warning about not handling the Cawbird URL while Hachyderm prints a warning about not handling its own URL.

File: lib/Base/System/ClientState.vala (hunk 5 above)
[the remark this answers is not in the document]
Comment (CodedOre) [reply lib/Base/System/ClientState.vala:368]: Alright, that makes sense. But then, you should also add an `unregister` method doing the same when an session is removed.

File: lib/Twitter/Utils/TextUtils.vala (hunk 3 above)
Comment (IBBoard) [reply lib/Twitter/Utils/TextUtils.vala:68]: This was definitely an ugly workaround that was abusing your string interpolation function to make it work 😁 But as I think I said in the commit message, it's required because some situations (bios on Twitter?) don't include an ID.

File: lib/Base/System/ClientState.vala (hunk 5 above)
[the remark this answers is not in the document]
Comment (IBBoard) [reply lib/Base/System/ClientState.vala:368]: Yes, true. Although I can't remember seeing anything bothering to unregister sessions 😄

File: src/Authentication/AuthView.vala (hunk 2 above)
Comment (IBBoard) [reply src/Authentication/AuthView.vala:143]: Ah yes. I remember pressing that now. Maybe I decided to skip that because it never worked for me and I assumed it was a "works on Twitter but not Mastodon" thing.

"Try again with authentication code" or similar wording might be clearer?

File: lib/Base/System/Client.vala (hunk 4 above)
[the remark this answers is not in the document]
Comment (IBBoard) [reply lib/Base/System/Client.vala:105]: Tried swapping back to the constructor. I get `'Backend.Client' already contains a definition of '.new'` as an error.

But it looks like that might be VSCode thing. It builds okay with Ninja. And using the constructor works fine.

File: lib/Twitter/System/Session.vala (hunk 1 above)
Comment (IBBoard) [reply lib/Twitter/System/Session.vala:62]: I'm still trying to work out this new flow. Does that mean that Twitter doesn't actually give an access token, but gives us a refresh token?

And were you storing it in `access_token` because that's the variable that we have? And it's the token that we need to track because it grants access to stuff (even if it's not the actual "access token")? I think that was part of what confused me about your `refresh_access_token` calls when I'd looked at other branches.

Comment (IBBoard) [reply lib/Twitter/System/Session.vala:62]: I've tried putting the refresh lines into the `init_async` on the Session. But I get a segfault that appears to be from the refreshed date not having a timezone. Even though I don't do anything with the timestamps between the refresh and the point that it errors.

```
#0  0x00007f5330c47be4 in g_time_zone_get_offset () at /lib64/libglib-2.0.so.0
#1  0x00007f5330bfbf7a in g_date_time_difference () at /lib64/libglib-2.0.so.0
#2  0x00007f5330bfbff9 in g_date_time_compare () at /lib64/libglib-2.0.so.0
#3  0x00007f5330a87fc8 in  () at /lib64/librest-1.0.so.0
#4  0x00007f5330a83207 in  () at /lib64/librest-1.0.so.0
#5  0x00007f5330a84637 in rest_proxy_call_invoke_async () at /lib64/librest-1.0.so.0
#6  0x00007f5330f9262b in backend_server_call_co (_data_=0x558f63597d30) at ../lib/Base/System/Server.vala:76
#7  0x00007f5330f92489 in backend_server_call (self=0x558f5f85bea0, call=0x558f638f58a0, cancellable=0x0, _callback_=0x7f5330fb35d0 <backend_twitter_session_init_async_ready>, _user_data_=0x558f5f742c70)
    at ../lib/Base/System/Server.vala:38
#8  0x00007f5330fb3e97 in backend_twitter_session_real_init_async_co (_data_=0x558f5f742c70) at ../lib/Twitter/System/Session.vala:112
#9  0x00007f5330fb356a in backend_twitter_session_real_init_async (self=0x558f635a5200, io_priority=0, cancellable=0x0, _callback_=0x7f5330f93b78 <backend_session_from_data_ready>, _user_data_=0x558f6348db70)
    at ../lib/Twitter/System/Session.vala:29
```

Comment (IBBoard) [reply lib/Twitter/System/Session.vala]: Slightly concerned that the logged time is `2123-01-14 @ 38947:37:03`. That's using `%T`. Which should be `%H:%M:%S`. So I don't have a clue how it gets to 38947 as a number of hours!

**Comment by IBBoard** [comment]
I've made some changes and pushed them. The only thing that's not working now is creating a Twitter account. I seem to get a segfault because the timezone on the token's expiry time is null and the date comparison assumes it isn't. Given that it's all managed by `librest` at that point then I don't have a clue what's going on.

File: lib/Twitter/System/Session.vala (hunk 1 above)
Comment (CodedOre) [reply lib/Twitter/System/Session.vala]: IIRC, the old code in `main` worked that way:
After the authentication, Twitter gives us an refresh and an access token. The access token is used and saved as known. However, at the next run of the program, the saved access token is used as the refresh token to acquire a new access token.

To be honest, after a few months not working on NewCaw, even I don't really understand why and how this works...

File: src/Cawbird.vala [hunk 6]
```diff
@@ -94,13 +94,23 @@ public class Cawbird : Adw.Application {
    * Initialize the client and open the first window.
    */
   protected override void activate () {
+    Backend.Client client;
     // Initializes the backend client
-    var client = new Backend.Client (Config.PROJECT_NAME,
-                                     "https://github.com/CodedOre/NewCaw",
-                                     "cawbird://authenticate");
+    try {
+      client = new Backend.Client("id",
```
Comment (CodedOre) [inline src/Cawbird.vala:100]:
```suggestion
      client = new Backend.Client(Config.APPLICATION_ID,
```

This should use the right id. Otherwise `KeyStorage` will save the access tokens for an application named `id`.

File: src/Cawbird.vala [hunk 7]
```diff
@@ -94,13 +94,23 @@ public class Cawbird : Adw.Application {
...
+        Config.PROJECT_NAME,
+        "https://github.com/CodedOre/NewCaw",
+        "cawbird://authenticate");
+    } catch (Error e) {
+      critical (@"Failed to load program state: $(e.message)");
+      return;
+    }
```
Comment (CodedOre) [inline src/Cawbird.vala:107]: Is the try-catch still needed?

File: src/Cawbird.vala (hunk 6 above)
Comment (IBBoard) [reply src/Cawbird.vala:100]: Ooops. Not sure how I managed to get that in there. Good catch!

File: src/Cawbird.vala (hunk 7 above)
Comment (IBBoard) [reply src/Cawbird.vala:107]: Possibly not. But the compiler didn't tell me that it wasn't needed, so I left it there in case there was an exception that I was missing. I'll remove it and see how well it works.

**Comment by IBBoard** [comment]
Okay. That seems to be working now. Some of the bugs trashed my profiles a couple of times, so I know I can re-add Mastodon accounts. And I fixed #64 in the process.

Also stripped out Twitter because… well, what's the point in supporting it now?

**Comment by CodedOre** [comment]
If we drop the Twitter platform, we don't really need the structure of the library with Base module and extending platforms anymore, so I think we could restructure this a bit more than just to remove a platform. Unless we have a potential different platform for which we want to leave the possibility open.

**Comment by IBBoard** [comment]
I was just leaving the possibility open 🙂 Deleting Twitter is quick. Flattening everything just seemed like too much work.

**Review by CodedOre — APPROVED** [review]
Alright, I'll merge this branch now.

We might want to discuss if we want to keep the multiple platform option or clean up the library to solely support Mastodon.

**Comment by IBBoard** [comment]
I don't see any harm in keeping it for now. You never know, maybe Musk will get bored and Twitter will fix everything he broke. Or some ex-Twitter devs will create a Twitter-like app with a similar API. And it will keep us considering a more generic architecture.

I already saw someone on Mastodon suggesting a Twitter-compatibility layer so that all of those apps can port more easily! (Although I doubt it would work that easily because of some of the different design and architecture decisions).
